blockchain/headers.go

```diff
@@ -0,0 +1,188 @@
+	// ErrHeaderCache is headers cache error class.
+	ErrHeaderCache = errs.Class("HeadersCache")
+
+	mon = monkit.Package()
+)
+
+// Header holds ethereum blockchain block header indexed data.
+// No need to keep number as big.Int right now as block count on ethereum mainnet is far from overflowing int64 capacity.
+type Header struct {
+	Hash      Hash
+	Number    int64
+	Timestamp time.Time
+}
+
+// HeadersDB is ethereum blockchain block header indexed cache.
+//
+// architecture: Database
+type HeadersDB interface {
+	// Insert inserts new header to cache db.
+	Insert(ctx context.Context, hash Hash, number int64, timestamp time.Time) error
+	// Delete deletes header from db by hash.
+	Delete(ctx context.Context, hash Hash) error
+	// Get retrieves header by hash.
+	Get(ctx context.Context, hash Hash) (Header, error)
+	// GetByNumber retrieves header by number.
+	GetByNumber(ctx context.Context, number int64) (Header, error)
+	// After retrieves block header which block timestamp is after provided time.
+	After(ctx context.Context, t time.Time) (Header, error)
+	// List retrieves all headers stored in cache db.
+	List(ctx context.Context) ([]Header, error)
+}
+
+// HeadersCache is a blockchain block header cache.
+type HeadersCache struct {
+	log *zap.Logger
+	db  HeadersDB
+}
+
+// NewHeadersCache creates new header cache.
+func NewHeadersCache(log *zap.Logger, db HeadersDB) *HeadersCache {
+	return &HeadersCache{
+		log: log,
+		db:  db,
+	}
+}
+
+// Get retrieves block header from cache by its hash.
+func (headersCache *HeadersCache) Get(ctx context.Context, hash Hash) (Header, bool, error) {
+	header, err := headersCache.db.Get(ctx, hash)
+	if err != nil {
+		if errs.Is(err, ErrNoHeader) {
+			return Header{}, false, nil
+		}
+
+		return Header{}, false, ErrHeaderCache.Wrap(err)
+	}
+
+	return header, true, nil
+}
+
+// HeaderSearch is block header search context.
+type HeaderSearch struct {
+	log         *zap.Logger
+	db          HeadersDB
+	endpoint    string
+	blockTime   time.Duration
+	threshold   time.Duration
+	batchLimit  int64
+	searchLimit int
+}
+
+// NewHeaderSearch creates new block header search context.
+func NewHeaderSearch(log *zap.Logger, endpoint string, blockTime, threshold time.Duration, batchLimit int64, searchLimit int) *HeaderSearch {
+	return &HeaderSearch{
+		log:         log,
+		endpoint:    endpoint,
+		blockTime:   blockTime,
+		threshold:   threshold,
+		batchLimit:  batchLimit,
+		searchLimit: searchLimit,
+	}
+}
+
+// After finds first block after specified timestamp and returns its header.
+func (headerSearch *HeaderSearch) After(ctx context.Context, t time.Time) (_ Header, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	c, err := rpc.DialContext(ctx, headerSearch.endpoint)
+	if err != nil {
+		return Header{}, ErrHeaderCache.Wrap(err)
+	}
+	defer c.Close()
+
+	client := ethclient.NewClient(c)
+	batchClient := newClient(c)
+
+	startBlock, err := client.HeaderByNumber(ctx, nil)
```

Comment:
Do you mean that If you give a time that is in the future and there the most recent block is in the past, right? If so yes it won't work but I can not imagine a case where such date will be provided

Reply:
Well, not necessarily future, but if the block header cache has not updated yet, and I give a time that is, say 15 seconds ago. But if you don't think it can happen, I'm fine with that.

Reply:
Yeah, I give it a thought and that makes sense actually. Will add it